Fixing user inputs for low, high in `make_tensor`

krshrimali opened the pull request:
**TODOs:**

* [x] Do not clamp inputs for low and high when given and valid.
* [x] Devise rules for modifying `low` and `high` when extremals/invalid values passed.
* [x] Testing with `test_references_numerics_hard` with the revised changes. _(I've tested locally, the changes will take place in a separate PR though after offline discussion with @mruberry)_
* [x] Revise comments/documentation for `make_tensor`

See https://github.com/pytorch/pytorch/issues/61758 for tracker issue.

cc: @mruberry @pmeier

krshrimali committed:
Init: do not clamp low and high values, hard assign ranges

krshrimali committed:
don't clamp low and high from the user, added bfloat16 to backward dtypes for 2 ops

krshrimali committed:
Tested OK, docs updated, remove OpInfo edits

krshrimali committed:
don't use _make_uniform, revert back + flake8 fixes

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into origin_dev/make_tensor_min_max

krshrimali committed:
Handle inf in low and high

krshrimali committed:
fix, use range_floats

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into origin_dev/make_tensor_min_max

krshrimali committed:
minor edits

krshrimali committed:
Update comments for -inf/inf input

krshrimali committed:
Merge branch 'master' into origin_dev/make_tensor_min_max

pmeier commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1987,22 +1989,26 @@ def make_tensor(size, device: torch.device, dtype: torch.dtype, *, low=None, hig
-        low = math.floor(0 if low is None else max(low, 0))
-        high = math.ceil(10 if high is None else min(high, 10))
+        ranges = (torch.iinfo(dtype).min, torch.iinfo(dtype).max)
+        low = math.floor(ranges[0] if low is float('-inf') else 0 if low is None or low < ranges[0] else low)
+        high = math.ceil(ranges[1] if high is float('inf') else 10 if high is None or high >= ranges[1] else high)
```
This is quite hard to read. I suggest you either put some parentheses around the second `if / else` or better yet write it as regular `if / elif`.

pmeier commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1973,10 +1973,12 @@ def make_tensor(size, device: torch.device, dtype: torch.dtype, *, low=None, hig
 
         By default, the tensor's values are in the range [-9, 9] for most dtypes. If low
-        and/or high are specified then the values will be in the range [max(-9, low), min(9, high)].
+        and/or high are specified then the values will be in the range [low, high). If low
+        and/or high are passed as -inf/inf (respectively), then the values will be minimum
+        and maximum limit of the respective dtype.
```
Not sure about this. I'm guessing this is only for BC? Could you give an example for such a case? If we actually want `inf` / `-inf` as part of the tensor, how would we achieve that?

krshrimali replied:
> I'm guessing this is only for BC? Could you give an example for such a case? 

The use case I encountered was from the OpInfos: there are ops which have domain listed as `(value, float('inf'))` (e.g: `acosh`: https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_methods_invocations.py#L4837). Though, it's expected that these domain values will be half open, which means `float('inf')` is going to be excluded, but still if you pass it in a function like `torch.distributions.uniform.Uniform` - it will give you infinite values:

```python
>>> torch.distributions.uniform.Uniform(low=2, high=float('inf')).sample((1, 2))
>>> tensor([[inf, inf]])
```

I'm not sure if it's a BC, since earlier it used to be: https://github.com/pytorch/pytorch/blob/2e49c5dc37bce5ce8ba463a59aacb5d3e4a638b6/torch/testing/_internal/common_utils.py#L1942:L1943. It's similar to what we are doing, except we use the dtype limits and not the default low and high values.

> If we actually want inf / -inf as part of the tensor, how would we achieve that?

From what I remember from our offline discussion, this does seem to be an interesting use-case where the user might want to include interesting values like nans/inf/-inf, but passing this as `low/high` doesn't seem to be the right way. We should discuss in the sync meeting, about having `include_inf` / `include_nan` like arguments OR `include_values` and allow users to specify the values they want to be included. Hopefully, these changes can be discussed and then implemented in the follow-up PRs.

Thanks for raising the question, hope this answers well. Let me know if you have any follow-up questions. @pmeier

pmeier replied:
Going along with the `acosh` example, wouldn't it be better to change the `OpInfo` to `domain=(1, None)`?

krshrimali replied:
Thinking of the reason why was domain set as `domain=(val, float('inf'))` for such ops (note that there are around 10 ops with such domains in the OpInfos):

A lot of mathematical resources mention these domains as `(val, inf)` or `(-inf, val)` and if this has to go to a user-facing API, passing `float('inf')` is still intuitive compared to passing `None`.

But I agree, that this can lead to some confusion, specially if we have something like `include_inf` arg and:

```python
# This is confusing since we pass inf to high but also ask to include_inf
make_tenosr(low=2, high=float('inf'), device, dtype, include_inf=True)

# This is not confusing
make_tensor(low=2, high=None, device, dtype, include_inf=True)
```

I agree that replacing the usages in the internals from `float('inf')` to `None` wheever reasonable, is a good idea but the first point about mathematical references mentioning `float('inf')` in the domains, is still an open question to be answered. Looking forward to hear your opinion. Thanks!

pmeier commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1987,22 +1989,26 @@ def make_tensor(size, device: torch.device, dtype: torch.dtype, *, low=None, hig
-        low = math.floor(0 if low is None else max(low, 0))
-        high = math.ceil(10 if high is None else min(high, 10))
+        ranges = (torch.iinfo(dtype).min, torch.iinfo(dtype).max)
+        low = math.floor(ranges[0] if low is float('-inf') else 0 if low is None or low < ranges[0] else low)
```
This silently clamps the lower bound to `0` in case the user passes a negative value. Shouldn't we simply raise an error, since the function is clearly used wrong?

krshrimali replied:
Good point, if we think of this going as a public facing API, raising an error sounds a good idea. Let's raise an error if `low < ranges[0]` (and low is not `float('-inf')`) or if `high > ranges[1]` for all the dtypes. What do you think?

pmeier replied:
The same of course applies to `high > ranges[1]` and also to all other branches.

krshrimali replied:
Okay, so we might to re-think on this point. Here is why:

`make_tensor` is internally used as: (example usage: https://github.com/pytorch/pytorch/blob/master/test/test_binary_ufuncs.py#L1674)

```
def some_function(device, some_dtype):
    make_tensor(size, device=device, dtype=some_dtype, low=-9, high=9)
```

low, high as `(-9, 9)` works for integral types like int, int64 but not for `uint8`. In these cases, the usage expects `make_tensor` to clamp the low and high values depending on the dtype.

The same is possible for public facing API, when they want to pass a generic (low, high) value but expect `make_tensor` to clamp for `uint8`.

```
# Example use case when you are making sure that integral types, floating types and complex types use [-1, 1) as the domain but uint8 takes [0, 1)
make_tensor(size, device, dtype, low=-1, high=1)
```

pmeier replied:
I see the use case, but I'm still against doing this silently and by default. As a reviewer if I read `make_tensor(..., low=-9, high=9)` I'm expecting the values to be in that range.

We could add a `clamp_to_dtype_range: bool = False` kwarg that enables this functionality. This should make it clear for the reviewer that some clamping is going on.

pmeier commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1987,22 +1989,26 @@ def make_tensor(size, device: torch.device, dtype: torch.dtype, *, low=None, hig
+        high = ranges_floats[1] if high is float('inf') else 9 if high is None or high >= ranges_floats[1] else high
         span = high - low
         # Windows doesn't support torch.rand(bfloat16) on CUDA
         if IS_WINDOWS and torch.device(device).type == 'cuda' and dtype is torch.bfloat16:
```
Since we are already touching `make_tensor`, is still true or can we remove this branch?

krshrimali replied:
Thanks for pointing this out, @pmeier - tested on my Windows machine that this isn't needed anymore. Have removed this condition.

pmeier commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2011,10 +2017,11 @@ def make_tensor(size, device: torch.device, dtype: torch.dtype, *, low=None, hig
             result = torch.rand(size, device=device, dtype=dtype) * span + low
     else:
         assert dtype in complex_types()
```
Couldn't we have complex dtypes as an extra `elif` clause? The author of this line clearly wanted to future proof `make_tensor` in case there is a new dtype that is not covered by the current branches. In that case the raised `AssertionError` is not really helpful. If we put complex in a separate branch, we could raise an expressive error in the `else` branch.

krshrimali replied:
Agreed, thanks for pointing this out. Done: https://github.com/pytorch/pytorch/pull/61108/commits/bafe03bdf63b33abd7565e9560bc4480fa8917de.

pmeier commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1987,22 +1989,26 @@ def make_tensor(size, device: torch.device, dtype: torch.dtype, *, low=None, hig
 
-    assert low is None or low < 9, "low value too high!"
-    assert high is None or high > -9, "high value too low!"
+    if low is not None and high is not None:
+        assert low <= high, "Expected low value to be lower than or equal to high, but found higher instead."
+        assert high >= low, "Expected high value to be higher than or equal to low, but found lower instead."
```
1. These are redundant since both statements test exactly the same thing.
2. For user facing errors we should prefer more expressive error types. Other than for testing functions `assert` should not be relied on during normal operation, since they can be disabled completely by the [`-O` flag](https://docs.python.org/3/using/cmdline.html#cmdoption-o).
3. We could make the error message more expressive by reporting the passed values back to the user.

```suggestion
    if low is not None and high is not None and low > high:
        raise ValueError(f"Expected low <= high, but got {low} > {high}")
```

pmeier reviewed (commented):
Thanks @krshrimali for the PR. Apart from my inline comments, I remember that discussed using `torch.distributions.Uniform` as the base sampler rather than `torch.rand`. What became of that plan?

krshrimali commented:
> Thanks @krshrimali for the PR. Apart from my inline comments, I remember that discussed using `torch.distributions.Uniform` as the base sampler rather than `torch.rand`. What became of that plan?

Thanks for the review, @pmeier - I have addressed a couple of your comments, will take a better look and update this PR with changes suggested. Regarding using: `torch.distriutions.uniform.Uniform` rather than `torch.rand`, I don't exactly remember why I reverted the commit. Will have to take a look again and see if it broke any relevant test. I'll get back to you on this after some testing. Thanks!

krshrimali committed:
Address reviews, add helper functions to keep it clean

krshrimali committed:
Revise doc for make_tensor

krshrimali committed:
Merge branch 'origin_dev/make_tensor_min_max' of https://github.com/krshrimali/pytorch into origin_dev/make_tensor_min_max

krshrimali committed:
From Philip's suggestion: Update torch/testing/_internal/common_utils.py

Co-authored-by: Philip Meier <github.pmeier@posteo.de>

krshrimali commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1973,10 +1973,16 @@ def make_tensor(size, device: torch.device, dtype: torch.dtype, *, low=None, hig
 
         By default, the tensor's values are in the range [-9, 9] for most dtypes. If low
-        and/or high are specified then the values will be in the range [max(-9, low), min(9, high)].
+        and/or high are specified and are within the dtype limits, then the values will be
+        in the range [low, high). However, if low and/or high are specified but are not in
+        dtype limits, then the dtype minimum and maximum limits are taken as low/high.
+        If low and/or high are passed as -inf/inf (respectively), then the values will be
+        minimum and maximum limit of the respective dtype, however if low is passed as inf
+        and high is passed as -inf, then a ValueError is raised (the same error is raised
+        if a nan value is passed).
```
In short, here is what I'm proposing (and what is happening): (pseudo code)

```python
if low/high is None:
    return (default_low, default_high)
else # if low/high are specified:
    if low in [nan, inf] or high in [nan, -inf]:
        # low can't be inf and high can't be -inf: hence two separate conditions
        raise ValueError(msg)
    elif low/high in range(dtype_min_range, dtype_max_range):
        return (low, high)
    elif (low == float('-inf') or low < dtype_min_range) or \
            (high == float('inf') or high > dtype_max_range):
        return (dtype_min_range, high) / (low, dtype_max_range) (respectively)
    elif low > dtype_max_range or high < dtype_min_range:
       raise ValueError(msg)
    else:
        return (low, high)
```

krshrimali commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1986,23 +1992,59 @@ def make_tensor(size, device: torch.device, dtype: torch.dtype, *, low=None, hig
         created tensor are replaced with an epsilon value if floating type, [`eps + `eps`.j] if
         complex type and 1 if integer/boolean type.
     """
+    def _modify_low_high(ranges, input_low, input_high, default_values, is_integral_dtype=False):
```
This helper function keeps prevents redundancies in if/else conditions below (when creating tensors). Since `make_tensor` is going to be looked at again in future PRs, I believe it's useful to add helper functions to keep the logic on modifying low/high values away from other logics.

krshrimali reviewed (commented):
This is ready for another round of review, unless and until there are unexpected test failures. A couple of comments still require attention, I'm looking at them. Thanks!

Leaving some sample outputs from modified `make_tensor`:

<details>
  <summary>Click to expand!</summary>

```python
>>> make_tensor((1, 2), device='cpu', dtype=torch.float32, low=-9, high=9)
tensor([[-8.4784, -1.7658]])
>>> make_tensor((1, 2), device='cpu', dtype=torch.uint8, low=-9, high=9)
tensor([[3, 2]], dtype=torch.uint8)
>>> make_tensor((1, 2), device='cpu', dtype=torch.uint8, low=-9, high=None)
tensor([[6, 4]], dtype=torch.uint8)
>>> make_tensor((1, 2), device='cpu', dtype=torch.uint8, low=None, high=None)
tensor([[2, 5]], dtype=torch.uint8)
>>> make_tensor((1, 2), device='cpu', dtype=torch.float32, low=None, high=None)
tensor([[ 3.1285, -3.0337]])
>>> make_tensor((1, 10), device='cpu', dtype=torch.float32, low=None, high=None)
tensor([[ 5.1067,  1.1351,  4.9473,  5.7744, -3.9730,  3.2708, -3.8939,  2.8211,
         -4.7024,  4.1631]])
>>> make_tensor((1, 10), device='cpu', dtype=torch.float32, low=None, high=float('inf'))
tensor([[2.0459e+38, 1.0356e+38, 8.6688e+37, 2.1416e+38, 3.2889e+38, 2.5179e+38,
         1.5371e+38, 1.6188e+38, 2.6685e+38, 5.1892e+37]])
>>> make_tensor((1, 2), device='cpu', dtype=torch.float32, low=float('inf'), high=float('inf'))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/krshrimali/pytorch/torch/testing/_internal/common_utils.py", line 2047, in make_tensor
    low, high = _modify_low_high(ranges_floats, low, high, default_values=(-9, 9))
  File "/home/krshrimali/pytorch/torch/testing/_internal/common_utils.py", line 2028, in _modify_low_high
    low = _for_val('low')
  File "/home/krshrimali/pytorch/torch/testing/_internal/common_utils.py", line 2008, in _for_val
    raise ValueError(f"Found invalid value {val} for {inp}")
ValueError: Found invalid value inf for low
>>> make_tensor((1,2), 'cpu', torch.float32, low=float('nan'), high=float('nan'))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/krshrimali/pytorch/torch/testing/_internal/common_utils.py", line 2048, in make_tensor
    low, high = _modify_low_high(ranges_floats, low, high, default_values=(-9, 9))
  File "/home/krshrimali/pytorch/torch/testing/_internal/common_utils.py", line 2029, in _modify_low_high
    low = _for_val('low')
  File "/home/krshrimali/pytorch/torch/testing/_internal/common_utils.py", line 2009, in _for_val
    raise ValueError(f"Found invalid value {val} for {inp}")
ValueError: Found invalid value nan for low
```
</details>
cc: @pmeier @mruberry

krshrimali committed:
Whitespace removed, linter check error

krshrimali committed:
Merge branch 'origin_dev/make_tensor_min_max' of https://github.com/krshrimali/pytorch into origin_dev/make_tensor_min_max

krshrimali committed:
Include nan comparison, break lines

krshrimali committed:
mvlgamma and log1p skips/fixes, TODOs added

krshrimali committed:
Fixing failures, ket's keep the max limit to 9 for internal testing?

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into origin_dev/make_tensor_min_max

krshrimali committed:
Replace 9 with None to prevent domain errors

krshrimali committed:
Raise ValueError if no valid dtype passed

krshrimali commented:
## Methods for generatic uniformly distributed tensors in the given range:

1. Using distribution function of PyTorch:
```python
low = torch.tensor(low_val, dtype=dtype, device=device)
high = torch.tensor(high_val, dtype=dtype, device=device)
size = (3, 3)

out = torch.distributions.uniform.Uniform(low, high).sample(size)
```

2. Using `torch.rand` and `span`:

```python
span = high - low
out = low + torch.rand(size, device=device, dtype=dtype) * span
```
## Comparison of distribution function of PyTorch vs `span` method: (CPU)

* **(1st Method) Time taken using uniform distribution of PyTorch**: 29.0877 microseconds
* **(2nd Method) Time taken using `torch.rand` and `span`**: 11.9051 microseconds

**Notes**:

1. The second method is ~59% faster than the first method.
2. In case the span is calculated between values and not tensors (i.e. `span = high_val - low_val` instead of `span = high - low`) then the time reduces to ~8.4517 microseconds.
3. The test is done for 1000 iterations (script is provided towards the end of this message).

## Comparison of distribution function of PyTorch vs `span` method: (CUDA)

* **(1st Method) Time taken using uniform distribution of PyTorch**: 78.5419 microseconds
* **(2nd Method) Time taken using `torch.rand` and `span`**: 34.6436 microseconds

**Notes**:

1. The second method is ~55% faster than the first method.
2. In case the span is calculated between values and not tensors (i.e. `span = high_val - low_val` instead of `span = high - low`) then the time reduces to ~18.8496 microseconds.
3. The test is done for 1000 iterations (script is provided towards the end of this message).

## Script

<details><summary>Click Here</summary>
<p>

```python
import torch
import time

# Decorator to record time over n_iters iterations for the function
def calc_time(n_iters):
    def f_wrapper(func):
        start = time.time()
        for _ in range(n_iters):
            func()
        end = time.time()
        print("Time taken: {:.8f} micro seconds".format((end-start)/n_iters * 1e6))
        return func()
    return f_wrapper

# Wraps the given value around torch.tensor for float32 dtype and cpu device
def tensor_wrapper(val):
    dtype, device = torch.float32, 'cpu'
    return torch.tensor(val, dtype=dtype, device=device)

low_val = torch.finfo(torch.float32).min
high_val = torch.finfo(torch.float32).max
low, high = tensor_wrapper(low_val), tensor_wrapper(high_val)
size = (3, 3)

@calc_time(1000)
def check_uniform():
    out = torch.distributions.uniform.Uniform(
            low=low/2, high=high/2
    ).sample(size)

@calc_time(1000)
def check_span():
    # We don't need to wrap the high/low values around tensors since we already pass dtype and device to torch.rand
    # For testing, you can uncomment the following 2 lines:
    # span = high/2 - low/2
    # out_old = torch.rand(size, device='cpu', dtype=torch.float32) * span + low
    out_old = torch.rand(size, device='cpu', dtype=torch.float32) * (high_val/2 - low_val/2) + low_val
```

</p>
</details>

## Observations

An interesting observation is that for the max and min limits of the given dtype, *both of the above methods will give you infinity values in the output*:

```python
>>> low = torch.tensor(torch.finfo(torch.float32).min, device='cpu', dtype=torch.float32)
>>> high = torch.tensor(torch.finfo(torch.float32).max, device='cpu', dtype=torch.float32)

>>> torch.distributions.uniform.Uniform(low, high).sample((3,))
tensor([inf, inf, inf])

>>> span = high - low
>>> span
tensor(inf)

>>> span = torch.finfo(torch.float32).max - torch.finfo(torch.float32).min
>>> span
6.805646932770577e+38
>>> low + torch.rand((3,), dtype=torch.float32, device='cpu') * span
tensor([inf, inf, inf])
```

For `make_tensor`, I hence will suggest to use: `high = <max_val>/2` and `low = <min_val>/2` so that span doesn't overflow.

Implementation wise, I think we should raise an error when the `range` overflows, this is because for `int32` if the range overflows, the output is `-1` because of the overflow:

```python
>>> high - low
tensor(-1, dtype=torch.int32)
```

NumPy raises an `OverflowError` which makes more sense to me: https://github.com/numpy/numpy/blob/2e5e6f6949af92b3f7abd99e558f1f6ae5be2185/numpy/random/mtrand.pyx#L1115:L1116.

cc: @mruberry @pmeier

krshrimali committed:
complex_types in elif for exclude_zeros; ranges to ranges/2 now

krshrimali committed:
Add whitespace around arithmetic operator

pmeier commented:
Thanks a lot @krshrimali for the investigation. I agree, if `torch.distributions` is slower and has the same flaws there is no reason for us to use it.

> Use max_limit/2 and min_limit/2 instead of max_limit and min_limit (since overflow leads to inf values in the output).

Since as far as I see the only reason is numerics, we can simply work around that:

```python
# inputs
dtype = torch.float32
low_val = torch.finfo(torch.float32).min
high_val = torch.finfo(torch.float32).max

# function body
span = high_val - low_val
dtype_max = torch.finfo(dtype).max
if span > dtype_max:
    factor = math.ceil(span / dtype_max)
    span /= factor
else:
    factor = 1

print((low_val / factor + torch.rand((3,), dtype=torch.float32, device='cpu') * span) * factor)
```

```
tensor([-2.2783e+37, -2.6669e+38,  1.7522e+38])
```

Toying with this lead me to find another problem: uniformly sampling with a large span will give us only large values. In fact about 90% of the values will have the same magnitude as the span. I think what we actually want is a uniform distribution over the different magnitudes, right? Otherwise intervals like `[1, inf)` make hardly sense, since the probability to get a value `< 10` is really tiny.

mruberry commented:
> Thanks a lot @krshrimali for the investigation. I agree, if `torch.distributions` is slower and has the same flaws there is no reason for us to use it.

Agreed! Let's stick with the simple approach.

> > Use max_limit/2 and min_limit/2 instead of max_limit and min_limit (since overflow leads to inf values in the output).

When would these limits come up? The default span is [-9, 9) or something similar. If the user inputs a span that is too big to represent we can throw a ValueError.

> Toying with this lead me to find another problem: uniformly sampling with a large span will give us only large values. In fact about 90% of the values will have the same magnitude as the span. I think what we actually want is a uniform distribution over the different magnitudes, right? Otherwise intervals like `[1, inf)` make hardly sense, since the probability to get a value `< 10` is really tiny.

Let's keep sampling simple and do our best approximation of a mathematically uniform sample. Sampling is a complicated topic and we could spend an infinite amount of time on it.

krshrimali committed:
Revise make_tensor for random tensor generation

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into origin_dev/make_tensor_min_max

krshrimali committed:
Fix for complex tensor generation

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into origin_dev/make_tensor_min_max

krshrimali commented:
> > > Use max_limit/2 and min_limit/2 instead of max_limit and min_limit (since overflow leads to inf values in the output).
> 
> When would these limits come up? The default span is [-9, 9) or something similar. If the user inputs a span that is too big to represent we can throw a ValueError.

These will only come when a user either passes the limits themselves (as low and high) or they pass (`float('-inf')` as low or `float('inf')` as high). This has been rectified though, thanks to @Lezcano for suggesting the workaround:

```python
x = torch.rand(...)
result = high * x + low * (1 - x)

# Instead of:
result = (high - low) * torch.rand(...) + low
```

I've updated the PR with the changes, waiting for the test cases to pass. The inline documentation for `make_tensor` needs updates (working on it).

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into origin_dev/make_tensor_min_max

krshrimali committed:
Documentation revised, keep default limits to -9 and 9

krshrimali committed:
change rtol for asin

krshrimali committed:
Minor correction with complex tensor gen

krshrimali committed:
Remove comments (not applicable anymore)

krshrimali committed:
Change tolerance for addbmm, failing on ROCm

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into origin_dev/make_tensor_min_max

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1,3 +1,4 @@
+""""""
```
What's up with this series of quotes?

krshrimali replied:
Oh, this got committed! I'm so sorry, will remove this.

krshrimali replied:
Fixed. Thanks!

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1939,38 +1943,74 @@ def make_tensor(size, device: torch.device, dtype: torch.dtype, *, low=None, hig
+                      'high': [input_high, default_values[1], float('-inf')]}
+
+        def _for_val(inp):
+            val, default_val, extremal_val = _vals_dict[inp][0], _vals_dict[inp][1], _vals_dict[inp][2]
```
Style nit: `= _vals_dict[inp]` should unpack the rhs for you

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1981,11 +2021,12 @@ def make_tensor(size, device: torch.device, dtype: torch.dtype, *, low=None, hig
             replace_with = torch.tensor(1, device=device, dtype=dtype)
         elif dtype in floating_types_and(torch.half, torch.bfloat16):
             replace_with = torch.tensor(torch.finfo(dtype).eps, device=device, dtype=dtype)
```
Would you update this line, too, to use `torch.finfo(dtype).tiny` instead of `.eps`?

krshrimali committed:
Address comments, waiting for tests to pass

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1939,38 +1943,74 @@ def make_tensor(size, device: torch.device, dtype: torch.dtype, *, low=None, hig
+        def _for_val(inp):
+            val, default_val, extremal_val = _vals_dict[inp][0], _vals_dict[inp][1], _vals_dict[inp][2]
+
+            if val != val or val == extremal_val:
```
This seems pretty convoluted. What about...

```
def _modify_low_high(low, high, lowest, highest, default_low, default_high, dtype):
  def clamp(a, l, h):
    return min(max(a, l), h)

  low = low if low is not None else default_low
  high = high if high is not None else default_high

  # Checks for error cases
  if low != low or high != high:
    raise ValueError("make_tensor: one of low or high was NaN!")
  if low > high:
    raise ValueError("make_tensor: low must be weakly less than high!")

  low = clamp(low, lowest, highest)
  high = clamp(high, lowest, highest)

  if dtype.is_floating_point or dtype.is_complex:
    return math.ceil(low), math.floor(high)
  
  return low, high
```

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1981,11 +2021,12 @@ def make_tensor(size, device: torch.device, dtype: torch.dtype, *, low=None, hig
             float_eps = torch.tensor(torch.finfo(float_dtype).eps, device=device, dtype=float_dtype)
             replace_with = torch.complex(float_eps, float_eps)
+        else:
+            raise ValueError(f"Invalid dtype passed, supported dtypes are: {get_all_dtypes()}")
```
I wouldn't make this change because we don't know what `get_all_dtypes()` will return in the future

mruberry reviewed (commented):
Overall this looks good, @krshrimali. I just made a couple suggestions and then we should wait for the tests

krshrimali committed:
Address reviews

krshrimali committed:
Minor fix with ceil and floor

krshrimali committed:
trying to fix asin failures

krshrimali committed:
Merge branch 'origin_dev/make_tensor_min_max' of https://github.com/krshrimali/pytorch into origin_dev/make_tensor_min_max

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into origin_dev/make_tensor_min_max

krshrimali committed:
rtol fix, trying again if the test passes

krshrimali committed:
Fix asin tolerance override decorator

krshrimali committed:
Add skips for linalg.det

krshrimali commented:
This is ready for a final review @mruberry - thanks for your patience and help. PTAL, whenever you find the time. :)

mruberry reviewed (approved):
Cool; finally a make_tensor() that respects the user's low and high!
